project-minimization-plugin/src/main/kotlin/org/plan/research/minimization/plugin/psi/PsiTrie.kt

```diff
@@ -0,0 +1,63 @@
+package org.plan.research.minimization.plugin.psi
+
+import org.plan.research.minimization.plugin.model.PsiWithBodyDDItem
+
+import com.intellij.openapi.application.readAction
+import com.intellij.psi.PsiElement
+import mu.KotlinLogging
+import org.jetbrains.kotlin.utils.addToStdlib.same
+
+/**
+ * A trie for Psi traversing. The implementation is mutable in terms of marking. However, new items could not be added.
+ * TODO!!!!!
+ */
+class PsiTrie private constructor() {
+    private val children: MutableMap<Int, PsiTrie> = mutableMapOf()
+    private var isMarked: Boolean = false
+    private val logger = KotlinLogging.logger {}
+    var hasMarkedElements: Boolean = false
+        private set
+
+    suspend fun processMarkedElements(element: PsiElement, processor: suspend (PsiElement) -> Unit) {
+        if (isMarked) {
+            if (logger.isTraceEnabled) {
+                // to preserve suspended context
+                logger.trace(
+                    readAction {
+                        "Processing marked element: ${element.textOffset} in ${element.containingFile.virtualFile.path}"
+                    },
+                )
+            }
+            processor(element)
+            return
+        }
+        for ((index, childTrie) in children) {
+            if (!childTrie.hasMarkedElements) {
+                continue
+            }
+            val childElement = readAction { element.children[index] }
+            childTrie.processMarkedElements(childElement, processor)
+        }
+    }
+
+    private fun add(item: PsiWithBodyDDItem, isFocused: Boolean, depth: Int = 0) {
+        if (depth == item.childrenPath.size) {
```

Comment:
Yea, it probably makes no sense, since it's not a classical recursion, but the inter-object one. Moreover, the caching is not the way, since for each `item` we query it once

Reply:
Ok, that's true :)